README.md
jQueryFileTree
==============
[![GitHub version](https://badge.fury.io/gh/jqueryfiletree%2Fjqueryfiletree.svg)](http://badge.fury.io/gh/jqueryfiletree%2Fjqueryfiletree)

##### NOTE:

__v2 is written in CoffeeScript and LESS. Now supports Bower and Gulp (for compiling).
View CHANGELOG.md to read more.__


DEMO
====
[http://jqueryfiletree.github.io/](http://jqueryfiletree.github.io/)


INSTALLING
==========
__Bower:__ `bower install jqueryfiletree --save`

__Manual:__ [Download ZIP](https://github.com/jqueryfiletree/jqueryfiletree/archive/master.zip)


ABOUT
=====

jQueryFileTree is a configurable, AJAX file browser plugin for jQuery. This repo is a continuation of unmaintained jQuery File Tree (12 April 2008) by Cory S.N. LaViska at ABeautifulSite.net

jQueryFileTree requires at least jQuery 1.2


FEATURES
========
* Produces valid, semantic XHTML
* Fully customizable via CSS
* Ability to style icons based on file extension
* Uses AJAX to fetch file information on the fly
* Easy to configure and implement
* Includes connector scripts for PHP and ASP.NET (C#)
* Supports custom connector scripts for extended functionality
* Customizable expand/collapse event
* Customizable expand/collapse speeds
* Supports easing functions
* Single- and multi-folder views
* Configurable load message
* Multi-select select with checkboxes
* Supports event listening on unique actions


CREATING A FILE TREE
====================

In it’s simplest form, you can create a file tree using the following code:

```javascript
	$(document).ready( function() {
		$('.class').fileTree({ root: '/some/folder/' }, function(file) {
			alert(file);
		});
	});
```

Where .class is the class of an empty DIV element that exists on your page. The file tree will automatically load when your page loads. Any DIV elements with this class will share the same file tree.


CONFIGURING THE FILE TREE
=========================
Parameters are passed as an object to the fileTree() function. Valid options include:

<table>
	<tr>
		<th>Parameter</th><th>Description</th><th>Default Value</th>
	</tr>
	<tr> <td>root</td><td>root folder to display</td><td>"/"</td> </tr>
	<tr> <td>script</td> <td>location of the serverside AJAX file to use</td> <td>"jqueryFileTree.php"</td> </tr>
	<tr> <td>folderEvent</td> <td>event to trigger expand/collapse</td> <td>"click"</td> </tr>
	<tr> <td>expandSpeed</td> <td>Speed to expand branches (in ms); use -1 for no animation</td> <td>500</td> </tr>
	<tr> <td>collapseSpeed</td> <td>Speed to collapse branches (in ms); use -1 for no animation</td> <td>500</td> </tr>
	<tr> <td>expandEasing \* </td> <td>Easing function to use on expand</td> <td>"swing"</td> </tr>
	<tr> <td>collapseEasing \* </td> <td>Easing function to use on collapse</td> <td>"swing"</td> </tr>
	<tr> <td>multiFolder</td> <td>Whether or not to limit the browser to one subfolder at a time</td> <td>true</td> </tr>
	<tr> <td>loadMessage</td> <td>Message to display while initial tree loads (can be HTML)</td> <td>"Loading..."</td> </tr>
	<tr> <td>errorMessage</td> <td>Message to display if unable to load tree</td> <td>"Unable to get file tree information"</td> </tr>
	<tr> <td>multiSelect</td> <td>Append checkbox to each line item to select more than one</td> <td>false</td> </tr>
	<tr> <td>onlyFolders</td> <td>Filter files and only return folders</td> <td>false</td> </tr>
	<tr> <td>onlyFiles</td> <td>Filter folders and only return files</td> <td>false</td> </tr>
</table>

\* _Anything other than 'swing' and 'linear' requires an external lib or script like [jQuery UI](http://jqueryui.com/) or [jquery.easing](https://github.com/gdsmith/jquery.easing/)_


There are many options available, which would look something like this:

```javascript
	$(document).ready( function() {
		$('.class').fileTree({
			root: '/some/folder/',
			script: 'jqueryFileTree.php',
			expandSpeed: 1000,
			collapseSpeed: 1000,
			multiFolder: false
		}, function(file) {
			alert(file);
		});
	});
```


STYLING THE FILE TREE
=====================
The file tree relies 100% on CSS for styling. Refer to jqueryFileTree.less to make any changes to the default styling.


CONNECTOR SCRIPTS
=================
jQueryFileTree comes with a handful of serverside connector scripts that are used to read the file system on your server and return data to the clientside script via AJAX. The default connector script is jqueryFileTree.php. You can use a connector script for another language by setting the script parameter to the location of the script you want to use (see Configuring the File Tree). Alternatively, you can build a custom connector script to extend the functionality of jQueryFileTree to better suit your needs (see Custom Connector Scripts).

Connector scripts for the following languages are provided:

* PHP by Cory LaViska (originally)
* ASP (VBS) by Chazzuka
* ASP.NET (C#) by Andrew Sweeny
* ColdFusion by Tjarko Rikkerink
* JSP by Joshua Gould
* Lasso by Marc Sabourdin
* Lasso by Jason Huck
* Perl by Oleg Burlaca
* Python/Django by Martin Skou
* Ruby by Erik Lax

(DAVE)
Note that all of the connector scripts have been left unmaintained outside of the PHP one in which I have updated (and will continue to do so). If you've improved or created a connector, feel free to create a pull request. *Use connector scripts as a starting point, but be mindful that often such (largely) unmaintained examples lack the security necessary for production.*


CUSTOM CONNECTOR SCRIPTS
========================
You can create a custom connector script to extend the functionality of the file tree. The easiest way to do this is probably by modifying one of the scripts supplied in the download. If you want to start from scratch, your script should accept one POST variable (dir) and output an unsorted list in the following format:

```html
	<ul class="jqueryFileTree">
		<li class="directory collapsed"><a href="#" rel="/this/folder/">Folder Name</a></li>
		(additional folders here)
		<li class="file ext_txt"><a href="#" rel="/this/folder/filename.txt">filename.txt</a></li>
		(additional files here)
	</ul>
```

Note that the corresponding file extension should be written as a class of the li element, prefixed with ext_. (The prefix is used to prevent invalid class names for file extensions that begin with non-alpha characters.)

Additionally you may choose to enable multi-select, which appends a checkbox to each item. Visible child elements will automatically be checked/unchecked along with the parent. Currently this is only supported in PHP; feel free to update other connectors to reflect the following format:

```html
	<ul class="jqueryFileTree">
		<li class="directory collapsed"><input type='checkbox' /><a href="#" rel="/this/folder/">Folder Name</a></li>
		(additional folders here)
		<li class="file ext_txt"><input type='checkbox' /><a href="#" rel="/this/folder/filename.txt">filename.txt</a></li>
		(additional files here)
	</ul>
```


EVENTS
=========================
jQueryFileTree now supports binding event listeners to the file tree element

```javascript
$('.filetree')
	.on('filetreeexpand', function (e, data)	{ console.log(data); })
	.on('filetreeexpanded', function (e, data)	{ console.log(data); })
	.on('filetreecollapsed', function (e, data)	{ console.log(data); })
	.on('filetreecollapse', function (e, data)	{ console.log(data); })
	.on('filetreeclicked', function(e, data)	{ console.log(data); });
```

All return the data object with the following properties

<table>
<tr>
	<th>Property</th><th>Value</th>
</tr>
<tr> <th>li</th><th>LI jQuery object</th> </tr>
<tr> <th>type</th><th>file | directory</th> </tr>
<tr> <th>value</th><th>name of the file or directory</th> </tr>
<tr> <th>rel</th><th>relative path to file or directory</th> </tr>
<tr> <th>container</th><th>container jQuery object (ie: `$('.filetree')`)</th> </tr>
</table>

Pretty much has the information you need, but I included the LI object anyways so you can easily get any other data you want with something like ``` data.li.prop('class') ```.


LICENSING & TERMS OF USE
========================
This plugin is dual-licensed under the GNU General Public License and the MIT License and is copyright 2008 A Beautiful Site, LLC.


TESTING
=======
In order to test, you'll need Bower and Gulp. Right now, I just have a manual browser demo to test functionality.
* In Terminal, go to the `/tests/manual/` folder and type `bower install` to set up the Bower assets.
* `gulp coffee` will compile `/src/coffeescript/jQueryFileTree.coffee` to JS, minify, and copy to `/dist/` as well as the testing folder (if testing is set up). A non-minified version is saved to `/src/` for debugging the output.
* `gulp less` will compile `/src/less/jQueryFileTree.less` to CSS, minify, and copy to `/dist/` as well as the testing folder (if testing is set up). A non-minified version is saved to `/src/` for debugging the output.
* `gulp` or `gulp default` will run `coffee` and `less` consecutively, then update the manual test folder.


SPECIAL THANKS
==============
A special thanks goes out to [FAMFAMFAM](http://www.famfamfam.com/) for their excellent Silk Icon Set.


CHANGELOG.md
### 2.0.2 - 08/13/2015
* Add javascript function as possible data source (#19)
* Some gulpfile.js modifications
* README updated

### 2.0.1 - 06/24/2015
* Fixing test CSS URL issue
* Moved repo to its own named organization and now using that name style throughout this project (jQueryFileTree)
* Updated Bower JSON files

### 2.0.0 - 06/24/2015
* Now written with LESS and CoffeeScript (core functionality remains essentially the same)
* Using a custom easing now checks that the easing function exists, else logs error and reverts to default easing (swing). ‘Swing’ now explicitly set as default rather than NULL.
* Creating Bower project - “jqueryfiletree”
* File structure updated to be more Bower-friendly (dist, src).
* Gulp support added for compiling LESS/CS
* Updating demo page to something way more expansive and informative. Moved into `/tests/manual/` and also utilizes Bower/Gulp.
* Removing event property `trigger` (type) because it’s just as easy to get with jQuery's `event.type`
* Adding event property `container` to easily find file tree when implementing more than one on a page
* Adding option `errorMessages` to allow users to customize error message if AJAX fails
* README updated


gulpfile.js
var gulp = require('gulp');
var coffee = require('gulp-coffee');
var gutil = require('gulp-util');
var uglify = require('gulp-uglify');
var rename = require("gulp-rename");
var minifyCss = require('gulp-minify-css');
var less = require('gulp-less');
var fs = require('fs');
//var changed = require('gulp-changed'); // doesn't seem to work; ALWAYS runs regardless

// compile LESS
gulp.task('less', function () {
    // jqueryfiletree assets
    gulp.src('src/less/*.less')
        //.pipe(changed('dist'), {extension: '.min.css'})
        .pipe(less())
        .pipe(gulp.dest('src'))
        .pipe(minifyCss())
        .pipe(rename( {extname: '.min.css'} ))
        .pipe(gulp.dest('dist'))
});

// compile coffeescript
gulp.task('coffee', function() {
    gulp.src('src/coffeescript/*.coffee')
        //.pipe(changed('dist'), {extension: '.min.js'})
        .pipe(coffee({bare: true}).on('error', gutil.log))
        .pipe(gulp.dest('src'))
        .pipe(uglify())
        .pipe(rename({ extname: '.min.js' }))
        .pipe(gulp.dest('dist'))
});

// 'gulp default' (or just 'gulp') will both build coffee/less, then reset test suite if applicable
gulp.task('default', ['coffee', 'less'], function(){
    // do if the test suite is set up
    attempt.test(function(){
        // copy images, connectors, and min files to tests/manual folder
        gulp.src(['dist/images/**', 'dist/connectors/**'], {base: 'dist'})
            //.pipe(changed('tests/manual'))
            .pipe(gulp.dest('tests/manual'));

        // compile skeleton & demo LESS files
        gulp.src(['tests/manual/bower_components/skeleton-less/less/*.less', 'tests/manual/css/less/*.less'])
            //.pipe(changed('tests/manual/css'), {extension: '.css'})
            .pipe(less())
            .pipe(gulp.dest('tests/manual/css'));

        // copy jQueryFileTree.min.css to test folder (if changed)
        gulp.src('dist/*.min.css')
            //.pipe(changed('tests/manual'))
            .pipe(gulp.dest('tests/manual'));

        // copy jQueryFileTree.min.js to test folder (if changed)
        gulp.src('dist/*.min.js')
            //.pipe(changed('tests/manual'))
            .pipe(gulp.dest('tests/manual'));
    });
});

// try/catch closure
var attempt = {
    // execute callback only if test suite is set up
    test: function(callback) {
        try {
            // Query if tests/bower_components is installed (else, that needs to be set up first)
            stats = fs.lstatSync('tests/manual/bower_components');

            // if bower is set up, the test suite is setup and we can run the test scripts
            if (stats.isDirectory()) {
                callback();
            }
        }
        catch (e) {
            // don't do anything with errors yet
        }
    }
}


jqueryFileTree.js
/**
 * jQuery File Tree Plugin
 *
 * @author - Cory S.N. LaViska - A Beautiful Site (http://abeautifulsite.net/) - 24 March 2008
 * @author - Dave Rogers - https://github.com/daverogers/jQueryFileTree
 *
 *
 * Usage: $('.fileTreeDemo').fileTree( options, callback )
 *
 * Options:  root           - root folder to display; default = /
 *           script         - location of the serverside AJAX file to use; default = jqueryFileTree.php
 *           folderEvent    - event to trigger expand/collapse; default = click
 *           expandSpeed    - default = 500 (ms); use -1 for no animation
 *           collapseSpeed  - default = 500 (ms); use -1 for no animation
 *           expandEasing   - easing function to use on expand (optional)
 *           collapseEasing - easing function to use on collapse (optional)
 *           multiFolder    - whether or not to limit the browser to one subfolder at a time
 *           loadMessage    - Message to display while initial tree loads (can be HTML)
 *           multiSelect    - append checkbox to each line item to select more than one
 *           onlyFolders    - show only folders
 *
 *
 * TERMS OF USE
 *
 * This plugin is dual-licensed under the GNU General Public License and the MIT License and
 * is copyright 2008 A Beautiful Site, LLC.
 */

if(jQuery) (function($){

	$.extend($.fn, {
		fileTree: function(options, file) {
			// Default options
			if( options.root			=== undefined ) options.root			= '/';
			if( options.script			=== undefined ) options.script			= '/files/filetree';
			if( options.folderEvent		=== undefined ) options.folderEvent		= 'click';
			if( options.expandSpeed		=== undefined ) options.expandSpeed		= 500;
			if( options.collapseSpeed	=== undefined ) options.collapseSpeed	= 500;
			if( options.expandEasing	=== undefined ) options.expandEasing	= null;
			if( options.collapseEasing	=== undefined ) options.collapseEasing	= null;
			if( options.multiFolder		=== undefined ) options.multiFolder		= true;
			if( options.loadMessage		=== undefined ) options.loadMessage		= 'Loading...';
			if( options.multiSelect		=== undefined ) options.multiSelect		= false;
			if( options.onlyFolders		=== undefined ) options.onlyFolders		= false;
			if( options.onlyFiles		=== undefined ) options.onlyFiles		= false;

			// internal vars
			var jqft = {
				initiatorClass: null
			}

			$(this).each( function() {

				function showTree(element, dir) {
					$(element).addClass('wait');
					$(".jqueryFileTree.start").remove();
					$.post(options.script,
					{
						dir: dir,
						onlyFolders: options.onlyFolders,
						onlyFiles: options.onlyFiles,
						multiSelect: options.multiSelect
					})
					.done(function(data){
						$(element).find('.start').html('');
						$(element).removeClass('wait').append(data);
						if( options.root == dir ) $(element).find('UL:hidden').show(); else $(element).find('UL:hidden').slideDown({ duration: options.expandSpeed, easing: options.expandEasing });
						bindTree(element);

						// if multiselect is on and the parent folder is selected, propagate check to child elements
						var li = $('[rel="'+decodeURIComponent(dir)+'"]').parent();
						if( options.multiSelect && li.children('input').is(':checked') ) {
							li.find('ul li input').each(function() {
								$(this).prop('checked', true);
								$(this).parent().addClass('selected');
							});
						}

						_trigger($(this), 'filetreeexpanded', data);
					})
					.fail(function(){
						$(element).find('.start').html('');
						$(element).removeClass('wait').append("<li>Unable to get file tree information</li>");
					});
				}

				function bindTree(element) {
					$(element).find('LI A').on(options.folderEvent, function() {
						// set up data object to send back via trigger
						var data = {};
						data.li = $(this).closest('li');
						data.type = ( data.li.hasClass('directory') ? 'directory' : 'file' );
						data.value	= $(this).text();
						data.rel	= $(this).prop('rel');

						if( $(this).parent().hasClass('directory') ) {
							if( $(this).parent().hasClass('collapsed') ) {
								// Expand
								_trigger($(this), 'filetreeexpand', data);

								if( !options.multiFolder ) {
									$(this).parent().parent().find('UL').slideUp({ duration: options.collapseSpeed, easing: options.collapseEasing });
									$(this).parent().parent().find('LI.directory').removeClass('expanded').addClass('collapsed');
								}

								$(this).parent().removeClass('collapsed').addClass('expanded');
								$(this).parent().find('UL').remove(); // cleanup
								showTree( $(this).parent(), encodeURIComponent($(this).attr('rel').match( /.*\// )) );
							} else {
								// Collapse
								_trigger($(this), 'filetreecollapse', data);

								$(this).parent().find('UL').slideUp({ duration: options.collapseSpeed, easing: options.collapseEasing });
								$(this).parent().removeClass('expanded').addClass('collapsed');

								_trigger($(this), 'filetreecollapsed', data);
							}
						} else {
							// this is a file click, return file information
							file($(this).attr('rel'));

							if( !options.multiSelect ) {
								// remove "selected" class if set, then append class to currently selected file
								$('div[class="'+jqft.initiatorClass+'"]').find('li').removeClass('selected');
								$(this).parent().addClass('selected');
							} else {
								// since it's multiselect, more than one element can have the 'selected' class
								if( $(this).parent().find('input').is(':checked') ) {
									$(this).parent().find('input').prop('checked', false);
									$(this).parent().removeClass('selected');
								} else
								{
									$(this).parent().find('input').prop('checked', true);
									$(this).parent().addClass('selected');
								}
							}

							_trigger($(this), 'filetreeclicked', data);
						}
						return false;
					});
					// Prevent A from triggering the # on non-click events
					if( options.folderEvent.toLowerCase != 'click' ) $(element).find('LI A').on('click', function() { return false; });
				}

				// Loading message
				$(this).html('<ul class="jqueryFileTree start"><li class="wait">' + options.loadMessage + '<li></ul>');

				// store the intiator class
				jqft.initiatorClass = $(this).attr('class');

				// Get the initial file list
				showTree( $(this), escape(options.root) );

				// wrapper to append trigger type to data
				function _trigger(element, eventType, data) {
					data.trigger = eventType;
					element.trigger(eventType, data);
				}

				// checkbox event (multiSelect)
				$(this).on('change', 'input:checkbox' , function(){
					var data = {};
					data.li		= $(this).closest('li');
					data.type	= ( data.li.hasClass('directory') ? 'directory' : 'file' );
					data.value	= data.li.children('a').text();
					data.rel	= data.li.children('a').prop('rel');

					// propagate check status to (visible) child checkboxes
					data.li.find('input:checkbox').prop( 'checked', $(this).prop('checked') );

					// set trigger
					if( $(this).prop('checked') ) {
						data.li.addClass('selected');				// add 'selected' class
						data.li.find('li').addClass('selected');	// also add to children
						_trigger($(this), 'filetreechecked', data);
					} else {
						data.li.removeClass('selected');
						data.li.find('li').removeClass('selected');
						_trigger($(this), 'filetreeunchecked', data);
					}
				});
			});
		}
	});

})(jQuery);


connectors/jqueryFileTree.php
<?php
if( !array_key_exists('HTTP_REFERER', $_SERVER) ) exit('No direct script access allowed');

/**
 * jQuery File Tree PHP Connector
 *
 * Version 1.1.0
 *
 * @author - Cory S.N. LaViska A Beautiful Site (http://abeautifulsite.net/)
 * @author - Dave Rogers - https://github.com/daverogers/jQueryFileTree
 *
 * History:
 *
 * 1.1.1 - SECURITY: forcing root to prevent users from determining system's file structure (per DaveBrad)
 * 1.1.0 - adding multiSelect (checkbox) support (08/22/2014)
 * 1.0.2 - fixes undefined 'dir' error - by itsyash (06/09/2014)
 * 1.0.1 - updated to work with foreign characters in directory/file names (12 April 2008)
 * 1.0.0 - released (24 March 2008)
 *
 * Output a list of files for jQuery File Tree
 */

/** 
 * filesystem root - USER needs to set this!
 * -> prevents debug users from exploring system's directory structure
 * ex: $root = $_SERVER['DOCUMENT_ROOT']; 
 */
$root = null;
if( !$root ) exit("ERROR: Root filesystem directory not set in jqueryFileTree.php");

$postDir = rawurldecode($root.(isset($_POST['dir']) ? $_POST['dir'] : null ));

// set checkbox if multiSelect set to true
$checkbox = ( isset($_POST['multiSelect']) && $_POST['multiSelect'] == 'true' ) ? "<input type='checkbox' />" : null;
$onlyFolders = ( isset($_POST['onlyFolders']) && $_POST['onlyFolders'] == 'true' ) ? true : false;
$onlyFiles = ( isset($_POST['onlyFiles']) && $_POST['onlyFiles'] == 'true' ) ? true : false;

if( file_exists($postDir) ) {

	$files		= scandir($postDir);
	$returnDir	= substr($postDir, strlen($root));

	natcasesort($files);

	if( count($files) > 2 ) { // The 2 accounts for . and ..

		echo "<ul class='jqueryFileTree'>";

		foreach( $files as $file ) {
			$htmlRel	= htmlentities($returnDir . $file);
			$htmlName	= htmlentities($file);
			$ext		= preg_replace('/^.*\./', '', $file);

			if( file_exists($postDir . $file) && $file != '.' && $file != '..' ) {
				if( is_dir($postDir . $file) && (!$onlyFiles || $onlyFolders) )
					echo "<li class='directory collapsed'>{$checkbox}<a href='#' rel='" .$htmlRel. "/'>" . $htmlName . "</a></li>";
				else if (!$onlyFolders || $onlyFiles)
					echo "<li class='file ext_{$ext}'>{$checkbox}<a href='#' rel='" . $htmlRel . "'>" . $htmlName . "</a></li>";
			}
		}

		echo "</ul>";
	}
}

?>


connectors/jqueryFileTree2.php
<?php

/**
 * jQuery File Tree PHP Connector
 *
 * Version 2.0.0
 *
 * @author - Cory S.N. LaViska A Beautiful Site (http://abeautifulsite.net/)
 * @author - Dave Rogers - https://github.com/daverogers/jQueryFileTree
 * @author - Lucian Nunu
 *
 * History:
 *
 * 2.0.0 - update to clean code (12/13/2014)
 * 1.1.0 - adding multiSelect (checkbox) support (08/22/2014)
 * 1.0.2 - fixes undefined 'dir' error - by itsyash (06/09/2014)
 * 1.0.1 - updated to work with foreign characters in directory/file names (12 April 2008)
 * 1.0.0 - released (24 March 2008)
 *
 * Output a list of files for jQuery File Tree
 */
$dir = urldecode((isset($_POST['dir']) ? $_POST['dir'] : null));

// set checkbox if multiSelect set to true
$checkbox = ( isset($_POST['multiSelect']) && $_POST['multiSelect'] == 'true' ) ? "<input type='checkbox' />" : null;

function scanRecursive($dir, $checkbox){
    
    if (file_exists($dir)) {
        $scan = array_diff(scandir($dir), array('..', '.'));

        if (count($scan)) {

            $directories = array();
            $files = array();

            foreach ($scan as $item) {
                if (is_dir($dir . $item)) {
                    $directories[] = $item;
                } else {
                    $files[] = $item;
                }
            }

            echo '<ul class="jqueryFileTree">';

            foreach ($directories as $file) {
                echo "<li class='directory collapsed'>{$checkbox}<a href='#' rel='" . htmlentities($dir . $file) . "/'>" . htmlentities($file) . "</a>";
                echo scanRecursive($dir . $file . '/', $checkbox);
                echo "</li>";
            }

            foreach ($files as $file) {
                $ext = preg_replace('/^.*\./', '', $file);
                echo "<li class='file ext_{$ext}'>{$checkbox}<a href='#' rel='" . htmlentities($dir . $file) . "'>" . htmlentities($file) . "</a></li>";
            }

            echo '</ul>';
        }
    }
}

scanRecursive($dir, $checkbox);




src/jQueryFileTree.js
/*
  * jQueryFileTree Plugin
  *
  * @author - Cory S.N. LaViska - A Beautiful Site (http://abeautifulsite.net/) - 24 March 2008
  * @author - Dave Rogers - https://github.com/daverogers/jQueryFileTree
  *
  * Usage: $('.fileTreeDemo').fileTree({ options }, callback )
  *
  * TERMS OF USE
  *
  * This plugin is dual-licensed under the GNU General Public License and the MIT License and
  * is copyright 2008 A Beautiful Site, LLC.
 */
(function($, window) {
  var FileTree;
  FileTree = (function() {
    function FileTree(el, args, callback) {
      var $el, defaults;
      $el = $(el);
      defaults = {
        root: '/',
        script: '/files/filetree',
        folderEvent: 'click',
        expandSpeed: 500,
        collapseSpeed: 500,
        expandEasing: 'swing',
        collapseEasing: 'swing',
        multiFolder: true,
        loadMessage: 'Loading...',
        errorMessage: 'Unable to get file tree information',
        multiSelect: false,
        onlyFolders: false,
        onlyFiles: false
      };
      this.jqft = {
        container: $el
      };
      this.options = $.extend(defaults, args);
      this.callback = callback;
      $el.html('<ul class="jqueryFileTree start"><li class="wait">' + this.options.loadMessage + '<li></ul>');
      this.showTree($el, escape(this.options.root));
    }

    FileTree.prototype.showTree = function(el, dir) {
      var $el, _this, data, handleFail, handleResult, options, result;
      $el = $(el);
      options = this.options;
      _this = this;
      $el.addClass('wait');
      $(".jqueryFileTree.start").remove();
      data = {
        dir: dir,
        onlyFolders: options.onlyFolders,
        onlyFiles: options.onlyFiles,
        multiSelect: options.multiSelect
      };
      handleResult = function(result) {
        var li;
        $el.find('.start').html('');
        $el.removeClass('wait').append(result);
        if (options.root === dir) {
          $el.find('UL:hidden').show(typeof callback !== "undefined" && callback !== null);
        } else {
          if (jQuery.easing[options.expandEasing] === void 0) {
            console.log('Easing library not loaded. Include jQueryUI or 3rd party lib.');
            options.expandEasing = 'swing';
          }
          $el.find('UL:hidden').slideDown({
            duration: options.expandSpeed,
            easing: options.expandEasing
          });
        }
        li = $('[rel="' + decodeURIComponent(dir) + '"]').parent();
        if (options.multiSelect && li.children('input').is(':checked')) {
          li.find('ul li input').each(function() {
            $(this).prop('checked', true);
            return $(this).parent().addClass('selected');
          });
        }
        return _this.bindTree($el);
      };
      handleFail = function() {
        $el.find('.start').html('');
        $el.removeClass('wait').append("<p>" + options.errorMessage + "</p>");
        return false;
      };
      if (typeof options.script === 'function') {
        result = options.script(data);
        if (typeof result === 'string' || result instanceof jQuery) {
          return handleResult(result);
        } else {
          return handleFail();
        }
      } else {
        return $.ajax({
          url: options.script,
          type: 'POST',
          dataType: 'HTML',
          data: data
        }).done(function(result) {
          return handleResult(result);
        }).fail(function() {
          return handleFail();
        });
      }
    };

    FileTree.prototype.bindTree = function(el) {
      var $el, _this, callback, jqft, options, relPattern;
      $el = $(el);
      options = this.options;
      jqft = this.jqft;
      _this = this;
      callback = this.callback;
      relPattern = /^\/.*\/$/i;
      $el.find('LI A').on(options.folderEvent, function() {
        var data, ref;
        data = {};
        data.li = $(this).closest('li');
        data.type = (ref = data.li.hasClass('directory')) != null ? ref : {
          'directory': 'file'
        };
        data.value = $(this).text();
        data.rel = $(this).prop('rel');
        data.container = jqft.container;
        if ($(this).parent().hasClass('directory')) {
          if ($(this).parent().hasClass('collapsed')) {
            _this._trigger($(this), 'filetreeexpand', data);
            if (!options.multiFolder) {
              $(this).parent().parent().find('UL').slideUp({
                duration: options.collapseSpeed,
                easing: options.collapseEasing
              });
              $(this).parent().parent().find('LI.directory').removeClass('expanded').addClass('collapsed');
            }
            $(this).parent().removeClass('collapsed').addClass('expanded');
            $(this).parent().find('UL').remove();
            _this.showTree($(this).parent(), $(this).attr('rel').match(relPattern)[0]);
            _this._trigger($(this), 'filetreeexpanded', data);
          } else {
            _this._trigger($(this), 'filetreecollapse', data);
            $(this).parent().find('UL').slideUp({
              duration: options.collapseSpeed,
              easing: options.collapseEasing
            });
            $(this).parent().removeClass('expanded').addClass('collapsed');
            _this._trigger($(this), 'filetreecollapsed', data);
          }
        } else {
          if (!options.multiSelect) {
            jqft.container.find('li').removeClass('selected');
            $(this).parent().addClass('selected');
          } else {
            if ($(this).parent().find('input').is(':checked')) {
              $(this).parent().find('input').prop('checked', false);
              $(this).parent().removeClass('selected');
            } else {
              $(this).parent().find('input').prop('checked', true);
              $(this).parent().addClass('selected');
            }
          }
          _this._trigger($(this), 'filetreeclicked', data);
          if (typeof callback === "function") {
            callback($(this).attr('rel'));
          }
        }
        return false;
      });
      if (options.folderEvent.toLowerCase !== 'click') {
        return $el.find('LI A').on('click', function() {
          return false;
        });
      }
    };

    FileTree.prototype._trigger = function(el, eventType, data) {
      var $el;
      $el = $(el);
      return $el.trigger(eventType, data);
    };

    return FileTree;

  })();
  return $.fn.extend({
    fileTree: function(args, callback) {
      return this.each(function() {
        var $this, data;
        $this = $(this);
        data = $this.data('fileTree');
        if (!data) {
          $this.data('fileTree', (data = new FileTree(this, args, callback)));
        }
        if (typeof args === 'string') {
          return data[option].apply(data);
        }
      });
    }
  });
})(window.jQuery, window);
